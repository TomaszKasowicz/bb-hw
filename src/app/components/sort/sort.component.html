<label>
  Sort By:
</label>
<button class="flex-grow-1" *ngFor="let sortColumn of sortColumns; index as i"
(click)="toggleSort(sortColumn.propertyName)"
[attr.data-testid]="'button-for-' + sortColumn.label"
>
  <div class="flex">
    <span>{{ sortColumn.label }}</span>    
    <div class="arrow" 
    [class.arrow-hidden]="sortColumn.propertyName !== state.property"
    [class.arrow-up]="state.ascending" 
    [class.arrow-down]="!state.ascending"
    [attr.data-testid]="'arrow-for-' + sortColumn.label"
    ></div>
    
  </div>
</button>