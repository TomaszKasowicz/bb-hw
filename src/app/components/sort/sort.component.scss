:host {
  display: flex;
}

label {
  margin-right: 1em;
}

button {
  background-color: var(--color-background-primary);
  border: 1px solid var(--color-line);

  & div {
    justify-content: center;
  }
}

.arrow {
  margin-left: 0.5em;
  width: 0;
  height: 0;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  flex-grow: 0;

  &-hidden {
    visibility: hidden;
  }

  &-up {
    border-bottom: 5px solid #999999;
  }

  &-down {
    border-top: 5px solid #999999;
  }
}
