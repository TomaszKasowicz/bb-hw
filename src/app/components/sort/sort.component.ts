import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { SortState } from '../../model/sort-state';

@Component({
  selector: 'app-sort',
  templateUrl: './sort.component.html',
  styleUrls: ['./sort.component.scss']
})
export class SortComponent implements OnInit {
  public sortColumns = [
    { label: 'DATE', propertyName: 'transactionDate' },
    { label: 'BENEFICIARY', propertyName: 'merchant' },
    { label: 'AMOUNT', propertyName: 'amount' }
  ];

  @Input()
  public state: SortState = {
    property: 'transactionDate',
    ascending: false
  };

  @Output()
  public sortChange = new EventEmitter<SortState>();

  constructor() {}

  ngOnInit(): void {
    this.sortChange.emit(this.state);
  }

  toggleSort(property: string) {
    const newState =
      property === this.state.property
        ? { ...this.state, ascending: !this.state.ascending }
        : { ...this.state, property };
    this.state = newState;
    console.log(newState);
    this.sortChange.emit(this.state);
  }
}
