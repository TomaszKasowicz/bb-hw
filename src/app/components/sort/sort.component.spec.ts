import { createComponentFactory } from '@ngneat/spectator/jest';
import { Spectator, byText, byTestId } from '@ngneat/spectator';

import { SortComponent } from './sort.component';

describe('SortComponent', () => {
  let spectator: Spectator<SortComponent>;
  const createComponent = createComponentFactory(SortComponent);

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('Should set sorting by Date by default', () => {
    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();
    const arrowForDateElem = spectator.query(byTestId('arrow-for-DATE'));
    expect(arrowForDateElem).not.toHaveClass(['arrow-hidden', 'arrow-up']);
    expect(arrowForDateElem).toHaveClass('arrow-down');
    expect(spectator.component.state).toEqual({
      property: 'transactionDate',
      ascending: false
    });
  });

  it('Arrow should only be visible for one active label', () => {
    spectator.detectChanges();
    const allArrowElements = spectator.queryAll('.arrow');
    const hiddenArrowElements = spectator.queryAll('.arrow-hidden');
    expect(allArrowElements.length).toEqual(hiddenArrowElements.length + 1);
  });

  it('Should toggle sorting by Date to descending when button clicked', () => {
    spectator.detectChanges();
    spectator.click(spectator.query(byTestId('button-for-DATE')) as any);
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
    const arrowForDateElem = spectator.query(byTestId('arrow-for-DATE'));
    expect(arrowForDateElem).not.toHaveClass(['arrow-hidden', 'arrow-down']);
    expect(arrowForDateElem).toHaveClass('arrow-up');
    expect(spectator.component.state).toEqual({
      property: 'transactionDate',
      ascending: true
    });
  });

  // todo: Convert this to it.each
  it('Should toggle sorting by Beneficiary when button clicked', () => {
    spectator.detectChanges();
    spectator.click(spectator.query(byTestId('button-for-BENEFICIARY')) as any);
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
    const arrowForBeneficiaryElem = spectator.query(
      byTestId('arrow-for-BENEFICIARY')
    );
    expect(arrowForBeneficiaryElem).not.toHaveClass([
      'arrow-hidden',
      'arrow-up'
    ]);
    expect(arrowForBeneficiaryElem).toHaveClass('arrow-down');
    expect(spectator.component.state).toEqual({
      property: 'merchant',
      ascending: false
    });
  });

  it('Should toggle sorting by Amount when button clicked', () => {
    spectator.detectChanges();
    spectator.click(spectator.query(byTestId('button-for-AMOUNT')) as any);
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
    const arrowForBeneficiaryElem = spectator.query(
      byTestId('arrow-for-AMOUNT')
    );
    expect(arrowForBeneficiaryElem).not.toHaveClass([
      'arrow-hidden',
      'arrow-up'
    ]);
    expect(arrowForBeneficiaryElem).toHaveClass('arrow-down');
    expect(spectator.component.state).toEqual({
      property: 'amount',
      ascending: false
    });
  });
});
