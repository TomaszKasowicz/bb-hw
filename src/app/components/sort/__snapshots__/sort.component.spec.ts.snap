// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortComponent Should set sorting by Date by default 1`] = `
<app-sort
  __ngContext__={[Function LRootView]}
  sortChange={[Function EventEmitter_]}
  sortColumns={[Function Array]}
  state={[Function Object]}
>
  <label>
     Sort By:
  
  </label><button
    class="flex-grow-1"
    data-testid="button-for-DATE"
  >
    <div
      class="flex"
    >
      <span>
        DATE
      </span>
      <div
        class="arrow arrow-down"
        data-testid="arrow-for-DATE"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-BENEFICIARY"
  >
    <div
      class="flex"
    >
      <span>
        BENEFICIARY
      </span>
      <div
        class="arrow arrow-hidden arrow-down"
        data-testid="arrow-for-BENEFICIARY"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-AMOUNT"
  >
    <div
      class="flex"
    >
      <span>
        AMOUNT
      </span>
      <div
        class="arrow arrow-hidden arrow-down"
        data-testid="arrow-for-AMOUNT"
      />
    </div>
  </button>
</app-sort>
`;

exports[`SortComponent Should toggle sorting by Amount when button clicked 1`] = `
<app-sort
  __ngContext__={[Function LRootView]}
  sortChange={[Function EventEmitter_]}
  sortColumns={[Function Array]}
  state={[Function Object]}
>
  <label>
     Sort By:
  
  </label><button
    class="flex-grow-1"
    data-testid="button-for-DATE"
  >
    <div
      class="flex"
    >
      <span>
        DATE
      </span>
      <div
        class="arrow arrow-down arrow-hidden"
        data-testid="arrow-for-DATE"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-BENEFICIARY"
  >
    <div
      class="flex"
    >
      <span>
        BENEFICIARY
      </span>
      <div
        class="arrow arrow-hidden arrow-down"
        data-testid="arrow-for-BENEFICIARY"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-AMOUNT"
  >
    <div
      class="flex"
    >
      <span>
        AMOUNT
      </span>
      <div
        class="arrow arrow-down"
        data-testid="arrow-for-AMOUNT"
      />
    </div>
  </button>
</app-sort>
`;

exports[`SortComponent Should toggle sorting by Beneficiary when button clicked 1`] = `
<app-sort
  __ngContext__={[Function LRootView]}
  sortChange={[Function EventEmitter_]}
  sortColumns={[Function Array]}
  state={[Function Object]}
>
  <label>
     Sort By:
  
  </label><button
    class="flex-grow-1"
    data-testid="button-for-DATE"
  >
    <div
      class="flex"
    >
      <span>
        DATE
      </span>
      <div
        class="arrow arrow-down arrow-hidden"
        data-testid="arrow-for-DATE"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-BENEFICIARY"
  >
    <div
      class="flex"
    >
      <span>
        BENEFICIARY
      </span>
      <div
        class="arrow arrow-down"
        data-testid="arrow-for-BENEFICIARY"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-AMOUNT"
  >
    <div
      class="flex"
    >
      <span>
        AMOUNT
      </span>
      <div
        class="arrow arrow-hidden arrow-down"
        data-testid="arrow-for-AMOUNT"
      />
    </div>
  </button>
</app-sort>
`;

exports[`SortComponent Should toggle sorting by Date to descending when button clicked 1`] = `
<app-sort
  __ngContext__={[Function LRootView]}
  sortChange={[Function EventEmitter_]}
  sortColumns={[Function Array]}
  state={[Function Object]}
>
  <label>
     Sort By:
  
  </label><button
    class="flex-grow-1"
    data-testid="button-for-DATE"
  >
    <div
      class="flex"
    >
      <span>
        DATE
      </span>
      <div
        class="arrow arrow-up"
        data-testid="arrow-for-DATE"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-BENEFICIARY"
  >
    <div
      class="flex"
    >
      <span>
        BENEFICIARY
      </span>
      <div
        class="arrow arrow-hidden arrow-up"
        data-testid="arrow-for-BENEFICIARY"
      />
    </div>
  </button><button
    class="flex-grow-1"
    data-testid="button-for-AMOUNT"
  >
    <div
      class="flex"
    >
      <span>
        AMOUNT
      </span>
      <div
        class="arrow arrow-hidden arrow-up"
        data-testid="arrow-for-AMOUNT"
      />
    </div>
  </button>
</app-sort>
`;
